pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,459 @@
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/common"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+const (
+	machineUIDTag       = "kubermatic-machine-controller:machine-uid"
+	defaultBillingCycle = "hourly"
+)
+
+// New returns a Packet provider
+func New(configVarResolver *providerconfig.ConfigVarResolver) cloud.Provider {
+	return &provider{configVarResolver: configVarResolver}
+}
+
+type RawConfig struct {
+	APIKey       providerconfig.ConfigVarString   `json:"apiKey"`
+	ProjectID    providerconfig.ConfigVarString   `json:"projectID"`
+	BillingCycle providerconfig.ConfigVarString   `json:"billingCycle"`
+	InstanceType providerconfig.ConfigVarString   `json:"instanceType"`
+	Facilities   []providerconfig.ConfigVarString `json:"facilities"`
+	OS           providerconfig.ConfigVarString   `json:"os"`
+	Tags         []providerconfig.ConfigVarString `json:"tags"`
+}
+
+type Config struct {
+	APIKey       string
+	ProjectID    string
+	BillingCycle string
+	InstanceType string
+	Facilities   []string
+	OS           string
+	Tags         []string
+}
+
+type provider struct {
+	configVarResolver *providerconfig.ConfigVarResolver
+}
+
+func (p *provider) getConfig(s v1alpha1.ProviderSpec) (*Config, *RawConfig, *providerconfig.Config, error) {
+	if s.Value == nil {
+		return nil, nil, nil, fmt.Errorf("machine.spec.providerconfig.value is nil")
+	}
+	pconfig := providerconfig.Config{}
+	err := json.Unmarshal(s.Value.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	rawConfig := RawConfig{}
+	if err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig); err != nil {
+		return nil, nil, nil, err
+	}
+
+	c := Config{}
+	c.APIKey, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.APIKey, "PACKET_API_KEY")
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"apiKey\" field, error = %v", err)
+	}
+	c.ProjectID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.ProjectID)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"projectID\" field, error = %v", err)
+	}
+	c.InstanceType, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.InstanceType)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"instanceType\" field, error = %v", err)
+	}
+	c.BillingCycle, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.BillingCycle)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"billingCycle\" field, error = %v", err)
+	}
+	if c.BillingCycle == "" {
+		c.BillingCycle = defaultBillingCycle
+	}
+	for i, tag := range rawConfig.Tags {
+		tagValue, err := p.configVarResolver.GetConfigVarStringValue(tag)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to get the %d value of \"tags\" field, error = %v", i, err)
```

Comment:
A potential error message would look like:
`failed to get the 0 value of "tags" field, error = something really bad happened`

Can we make this more human readable?

Reply:
Happy to take a better suggestion. I took this straight from the Hetzner example, [here](https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/hetzner/provider.go#L85)

Reply:
What do you think about `failed to load the value for the Tag at index %d, error = %v` ?

Reply:
Did something almost like that. Should be good enough.